File: src/telecom/repositories/subscriber-repository.ts
import type { TelecomSubscriber } from '../entities/telecom-subscriber.js';

/**
 * Subscriber repository contract for the telecom simulation (modules 03/04).
 *
 * Two adapters ship with TURANT:
 *
 *   - InMemorySubscriberRepository : the sim's in-process store (tests / local
 *     dev, no database needed).
 *   - PostgresSubscriberRepository : real PostgreSQL tables seeded by the sim
 *     seeder (the 1K → 300M path), with configurable table/column names so it
 *     can later point at the real C-DOT subscriber schema without code changes.
 *
 * The real C-DOT repository that modules 03/04 will eventually use must
 * implement the same interface; nothing else in the pipeline changes.
 */

/** The shape the matcher actually needs per matched cell. */
export interface SubscriberRow {
  imsi: string;
  msisdn: string;
  cellId: string;
  towerId?: string;
  technology?: string;
  status?: string;
  lastSeen?: Date;
}

export interface FindByCellIdsOptions {
  /** Safety cap per lookup. */
  limit?: number;
}

export interface SubscriberRepository {
  /** Stable name for logs/audit, e.g. "telecom-sim:memory" | "telecom-sim:postgres". */
  readonly name: string;
  /** Total subscribers currently stored (seeding idempotency/resume). */
  count(): Promise<number>;
  /** Insert or refresh a batch of subscribers (idempotent by IMSI). */
  upsertSubscribers(rows: readonly TelecomSubscriber[]): Promise<number>;
  /** Subscribers currently attached to any of the given cell ids. */
  findByCellIds(cellIds: readonly string[], options?: FindByCellIdsOptions): Promise<SubscriberRow[]>;
  /** Full records for a list of MSISDNs (module 04 lookup path). */
  findByMsisdns(msisdns: readonly string[], options?: FindByCellIdsOptions): Promise<TelecomSubscriber[]>;
}

/** Thrown by the repository factory when the real C-DOT store is not configured. */
export class SubscriberRepositoryNotConfiguredError extends Error {
  constructor() {
    super('Real C-DOT Subscriber Repository Not Configured');
    this.name = 'SubscriberRepositoryNotConfiguredError';
  }
}

/** Base class for repository I/O failures. */
export class SubscriberRepositoryError extends Error {
  constructor(message: string, override readonly cause?: unknown) {
    super(message);
    this.name = 'SubscriberRepositoryError';
  }
}

File: src/main/java/com/turant/subscriber/SubscriberRepository.java
package com.turant.subscriber;

import com.turant.types.subscriber.Subscriber;

import java.util.List;
import java.util.concurrent.CompletableFuture;

/**
 * Subscriber repository contract for modules 03/04.
 * 
 * Provides access to subscriber data for matching against cell towers.
 * 
 * Migrated from TypeScript telecom/repositories/subscriber-repository.ts
 */
public interface SubscriberRepository {
    
    /**
     * Stable name for logs/audit, e.g. "telecom-sim:memory" | "telecom-sim:postgres"
     */
    String getName();
    
    /**
     * Total subscribers currently stored.
     */
    CompletableFuture<Long> count();
    
    /**
     * Subscribers currently attached to any of the given cell IDs.
     *
     * @param cellIds List of cell IDs to match
     * @param options Search options (limit)
     * @return List of subscriber rows
     */
    CompletableFuture<List<SubscriberRow>> findByCellIds(List<String> cellIds, FindByCellIdsOptions options);
    
    /**
     * Full records for a list of MSISDNs (lookup path).
     *
     * @param msisdns List of MSISDNs
     * @param options Search options
     * @return List of full subscriber records
     */
    CompletableFuture<List<Subscriber>> findByMsisdns(List<String> msisdns, FindByCellIdsOptions options);
    
    /**
     * The shape the matcher needs per matched cell.
     */
    record SubscriberRow(
        String imsi,
        String msisdn,
        String cellId,
        String towerId,
        String technology,
        String status,
        java.time.Instant lastSeen
    ) {}
    
    /**
     * Options for subscriber lookup.
     */
    class FindByCellIdsOptions {
        private Integer limit;
        
        public Integer getLimit() {
            return limit;
        }
        
        public FindByCellIdsOptions setLimit(Integer limit) {
            this.limit = limit;
            return this;
        }
        
        public static FindByCellIdsOptions defaults() {
            return new FindByCellIdsOptions();
        }
    }
}
